### DISCUSSION TITLE
[RFC] Generate ActiveAdmin like pages

### DISCUSSION THREAD

**Discussion by engelkes-finstreet** [blitz-js/blitz#1166] [RFCs]
One thing I love about Ruby on Rails is their ActiveAdmin. The ActiveAdmin is a separate part of the app which lets your admins manage all your models in a pretty simple way without writing too much boilerplate code.

### Proposal

We basically generate all mutations right now anyways with `CREATE, UPDATE and DELETE` and even create the pages. In my first try I added _CREATE / EDIT / DELETE_ buttons inside my App and only made them visible to admins. A cleaner approach is a completely separate section where all of this is possible.

Each App should get a folder structure like this `app/admin/pages` and inside this would be nearly an identical folder structure as we have it in our app right now `/users/pages/[userSlug]/edit` and `/users/pages/new`... you get the idea.
All of this could be wrapped in a separate Layout which is protected by the `useAuthorize` hook. So the hook is only needed in one place and not all over your app.

We could add a new `blitz generate admin` command which would create all the needed admin files.

One thing which I do not like about ActiveAdmin is that it is kind of static. Doing something more fancy was mostly a pain in the $%@. Generating an ActiveAdmin for BlitzJS we could generate all forms but customizing this would be way easier.

**Comment by Zeko369** [comment]
Hmm, if we exclude customizing, Prisma Studio behind auth is what you're looking for or did I miss your point (and that the whole idea of this is to make it customizable)

**Reply by engelkes-finstreet** [reply]
The customizable point is missing for me. I often have to create AdminSections which have the same Look & Feel of the App or give you more than just simple CRUD operations. Unfortunately Prisma Studio is not enough for my use cases.

**Comment by ntgussoni** [comment]
The DSL is what makes AA so good. With almost no code you get CRUD operations for all your models.
On the other hand the DSL also gets in your way when you try go to off the path. 

I guess the biggest challenge here is to make it as easy as possible without losing flexibility

**Comment by MrLeebo** [comment]
This sounds like it could be accomplished as an enhancement to the existing generate command. For instance, a `--layout` parameter; 

```bash
blitz generate pages admin/products --layout AdminLayout
```

then there wouldn't need to be a new command.

**Comment by svobik7** [comment]
I like the idea of having `admin` stuff separated inside its own `app/admin` folder. But one thing came to my mind “what if different _app, 404 or _document files are needed (e.g. for different UI/theme assets”?

How this would be solved?

**Comment by flybayer** [comment]
Thanks @engelkes-finstreet, we definitely need something like this!!

For most of my uses cases, I just want to deploy something basic like prisma studio along with my blitz app (single deployment with shared auth). Being able to generate and customize the code would be awesome addition when needed.

It seems this could be really cool as a blitz plugin. 

By default, just installing the plugin would not generate code into your project, but it would dynamically inject all the admin pages and crud into your app at build time. This way the admin section is ALWAYS up to date with your prisma schema without changing anything (same as rails active admin).

Then also have the ability to "eject" everything or certain parts into your app where you can customize.

**Reply by engelkes-finstreet** [reply]
This sounds like a good idea. Are there already some examples for a plugin / dynamically injecting code? 
I fear that I lack a lot of knowledge to implement it the way you suggested.

**Reply by flybayer** [reply]
We haven't done any work on this yet. And I definitely won't have time to focus on this until after 1.0.

However, you can certainly work on figuring out what the actual generated code looks like and go as far as you can on this. Then we can easily port that into a plugin once we add plugin support.

**Reply by engelkes-finstreet** [reply]
Totally understandable. I was looking into the server package yesterday and I think I will for now write a generator for me which creates the desired files which we can port to a plugin after 1.0

**Comment by cr101** [comment]
For our use case the Admin Dashboard is wrapped in the same Layout and the users who are members of an Organisation have edit rights to update the info for that Organisation.

We are building a company register web app where the details of every company are visible to the public and the directors and the founders of a company can login and change the details of that company.

**Reply by svobik7** [reply]
This is slightly different use case then what we do in our “non-blitz yet” project.

What you describes sounds like “client admin”. Users can manage their own entries.

In my use case I would create “main admin” where chosen users (administrators) have permission to manage all entries, all non-admin users, etc..

**Reply by flybayer** [reply]
I've started calling this role the "superadmin".

**Reply by cr101** [reply]
yes “client admin” is what I'm referring to. It's a common use case for web apps and some examples are Yelp, Zomato, Crunchbase, etc which allow the organization/store owners to update certain information.

**Reply by svobik7** [reply]
@cr101 we also use this kind of “client admin” and it is also wrapped in the same layout as whole client app. Eg. when user add a post on `pages/posts/add.tsx` then they could manage the same product on `pages/posts/[id]/edit.tsx`. That is good totally ok.

But then we also have “superadmin” (like @flybayer wrote) where our internal team superadmins are allowed to manage whole app, see graphs and stats which are private. This superadmin UI has different layout, css and RBAC but share some models logic.

So I am thinking about what would be the best approach here to minimalize code duplication.

**Comment by flybayer** [comment]
I just posted a guide on [How to Run Prisma Studio in Production on Render.com](https://github.com/blitz-js/blitz/wiki/How-to-Run-Prisma-Studio-in-Production-on-Render.com) to function as an admin dashboard.
